### PR TITLE
Add --rootmount={shared:private:slave} flags to set rootfs mount 

### PR DESCRIPTION
**Pull request by rootfs**
This is the docker part of https://github.com/docker/libcontainer/pull/632

It implements [a rootfs mount propagation flag](https://github.com/docker/libcontainer/pull/632#issuecomment-112177281)

### CODE REVIEW

**Comment by phemmer** [comment]
If i'm reading the linked issue correctly, this would close #10088

**Comment by rootfs** [comment]
@phemmer we indeed have similar use case.

**Comment by rhatdan** [comment]
We need to get a move on this,  This is critical to us for docker-1.8.  We are carrying a patch but k8s and openshift communities on non Red Hat Platforms need to get the rootfs to be slave not private.

**Comment by rhatdan** [comment]
@rootfs Could you make this patch match the one you have for runc.

**Comment by rootfs** [comment]
@rhatdan working on it, thanks.

**Comment by LK4D4** [comment]
runc PR opencontainers/runc#77

**Comment by timothysc** [comment]
@kelseyhightower, @thockin - can we unite and get some leverage on this one?

**Comment by rootfs** [comment]
thanks @timothysc. A docker prototype that included https://github.com/opencontainers/runc/pull/77  can be evaluated from [here](https://huaminchen.wordpress.com/2015/07/31/latest-status-on-docker-mount-propagation-development/)

**Comment by thockin** [comment]
My biggest fear here is that this is a *really* complicated semantic which is really useful in a small set of cases.  But I don't have a better answer - this is certainly better than using nsenter to jump between NSes.  I suspect only 2 of the 3 modes are actually useful, but it's hard to make a case to exclude 1/3 of a kernel feature.  Not sure how this applies to other Docker platforms...

TL;DR +1

**Comment by thockin** [comment]
@icecrime is there any philosophical objection to this?  It's unfortunately a bit complex, but your own example of FUSE daemons in containers is exactly right.

**Comment by childsb** [comment]
For many filesystems, FUSE daemons, run-running client baggage and configs are required on the host.

With this patch we could containerize all of that headache and really make filesystem access portable across platforms without the setup and daemon dogma on the host.  Imagine connecting to cinder shares on a host without installing complicated nova packages...  Its all in a container cleanly hidden away.  The host just sees a mountpoint.

HUGE +1 .. This would be an incredible feature.  I beg of the docker higher powers... We have lots of exciting guides, articles, and everything else boiling around this particular design.  Please merge!!!!

**Comment by jayunit100** [comment]
I guess we cant technically +1 cause its not mergeable, but  definetly +1 pending rebase, lets get this guy in.  the value of containerizing pluggable storage is a really important feature for the docker ecosystem as a whole.

**Comment by cpuguy83** [comment]
This is a super low-level control... but we do need something like this especially for containerizing volume plugins.
Since we can't really share mount namespaces between containers/host this seems to be the path forward for this kind of functionality.

I'd like the discuss the implications of this, though.
I would prefer to make sure people don't mount stuff without dong it in a volume. Basically prevent `docker run --rootmount=shared myimage bash -c 'mount <opts> /foo'` and then people consuming `/var/lib/docker/<graph dir>/<container id>/foo` directly.

So I guess the real issue is, can we solve a specific use case vs a generic option that will be abused.

**Comment by rootfs** [comment]
@cpuguy83 nice point. For your example, I think the way the patch is implemented doesn't allow people consuming directories under `/var/lib/docker/<graph dir>/<container id>/foo` directly. The mountpoint can only be created the host, because in the latest code, [the rootfs is marked private](https://github.com/opencontainers/runc/pull/77/files#diff-f41204a863eb2caa17fd9f68d119da63R412). So it is more like the following scenario:

`docker run --rootmount=shared myimage -v /:host bash -c 'mount <opts> /host/foo'`

**Comment by cpuguy83** [comment]
@rootfs what's the use-case for `slave`?

**Comment by rhatdan** [comment]
We are preparing a new pull request for this.  With everyone at Red Hat in agreement.  :^)

slave is for the autofs and k8s use case.  Basically you want to mount a volume into a container, and then any mounts on top of that volume will show up inside the container.

**Comment by cpuguy83** [comment]
@rhatdan Doesn't this already happen today, assuming the daemon isn't in it's own mount namespace?

**Comment by rootfs** [comment]
@cpuguy83 I don't think containers can see new mounts propagated from host, since the [rootfs propagation is private](https://github.com/docker/docker/blob/master/daemon/execdriver/driver_linux.go#L30)

**Comment by rhatdan** [comment]
@cpuguy83 No it does not since libcontainer reverted my patch.

libcontainer does a mount --make-rprivate /
during setup, which basically isolates the container from its parents mount namespace.

The change we are trying to get is basically 

mount --make-rslave /

Which fixes the use case.

We are also working on 

mount --make-rshared /

For the SPC use case of mount client commands wanting to mount file systems onto the parents mnt namespace.  (Cefs,Cluster, nfs for example)

**Comment by rootfs** [comment]
Closed, please comment on the new PR #15648
